futures-util/src/future/flatten_stream.rs

```diff
@@ -58,34 +61,22 @@ impl<Fut> Stream for FlattenStream<Fut>
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         loop {
             // safety: data is never moved via the resulting &mut reference
-            let stream = match &mut unsafe { Pin::get_unchecked_mut(self.as_mut()) }.state {
+            match &mut unsafe { Pin::get_unchecked_mut(self.as_mut()) }.state {
```

Comment:
To encapsulate this unsafety you may be able to add something like

```rust
impl<Fut: Future> State<Fut> {
    fn get_pin_mut(self: Pin<&mut Self>) -> State<Pin<&mut Fut>>
}
```

I think that might work since `<Fut as Future>::Output == <Pin<&mut Fut> as Future>::Output`, but `rustc` also might not be able to determine that I guess.

Reply:
Perhaps that doesn't work well.
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5149d533324f5fad6c4056b230f8117e

Reply:
Actually, that's a different issue than what I was afraid of, it's definitely not possible to have that API because of the `Stream` variant holding it by value.

You could split it into a generic `Either` like https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0c27eb07e5693dab683e855a30aa961b, but I'm not certain it's worth it.

Reply:
Another way I'm thinking is to use generics in `State`, not associated types.
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=36bf9cde7b51f3ac63355fb2fb3145a9